subprojects/publish/src/main/java/org/gradle/api/publish/internal/mapping/DefaultDependencyCoordinateResolverFactory.java

```diff
@@ -0,0 +1,159 @@
+    }
+
+    public DependencyResolvers getDependencyResolutionStrategy(SoftwareComponentVariant variant) {
+        Configuration configuration = null;
+
+        ImmutableAttributes attributes = ((AttributeContainerInternal) variant.getAttributes()).asImmutable();
+        VariantVersionMappingStrategyInternal versionMapping = versionMappingStrategy.findStrategyForVariant(attributes);
+
+        // Fallback to component coordinate mapping if variant mapping is not enabled
+        ComponentDependencyResolver componentResolver;
+        if (versionMapping.isEnabled()) {
+            if (versionMapping.getUserResolutionConfiguration() != null) {
+                configuration = versionMapping.getUserResolutionConfiguration();
+            } else if (versionMapping.getDefaultResolutionConfiguration() != null) {
+                configuration = versionMapping.getDefaultResolutionConfiguration();
+            }
+        }
+
+        if (configuration != null) {
+            componentResolver = new VersionMappingVariantDependencyResolver(projectDependencyResolver, configuration);
+        } else {
+            componentResolver = new MinimalComponentDependencyResolver(projectDependencyResolver);
+        }
+
+        return new DependencyResolvers(new VariantResolverAdapter(componentResolver), componentResolver);
+    }
+
+    private static class DependencyResolvers {
+
+        private final VariantDependencyResolver variantResolver;
+        private final ComponentDependencyResolver componentResolver;
+
+        public DependencyResolvers(VariantDependencyResolver variantResolver, ComponentDependencyResolver componentResolver) {
+            this.variantResolver = variantResolver;
+            this.componentResolver = componentResolver;
+        }
+    }
+
+    /**
+     * Adapts a {@link ComponentDependencyResolver} to a {@link VariantDependencyResolver}
+     * by returning component-precision coordinates.
+     */
+    private static class VariantResolverAdapter implements VariantDependencyResolver {
+
+        ComponentDependencyResolver delegate;
+
+        public VariantResolverAdapter(ComponentDependencyResolver delegate) {
+            this.delegate = delegate;
+        }
+
+        @Nullable
+        @Override
+        public ResolvedCoordinates resolveVariantCoordinates(ExternalDependency dependency, VariantWarningCollector warnings) {
+            return delegate.resolveComponentCoordinates(dependency);
+        }
+
+        @Override
+        public ResolvedCoordinates resolveVariantCoordinates(ProjectDependency dependency, VariantWarningCollector warnings) {
+            return delegate.resolveComponentCoordinates(dependency);
+        }
+    }
+
+    /**
+     * A {@link ComponentDependencyResolver} which does not depend on resolving a dependency graph.
+     */
+    private static class MinimalComponentDependencyResolver implements ComponentDependencyResolver {
```

Comment:
💭 Instead of `Minimal` how about `ProjectOnly` as the prefix on this type?

Also, as this is new code, I'd strongly consider returning `Optional` and using `empty()` to signify no result, instead of `null`.  We should minimize the use of `null` returns when possible.

Reply:
I would prefer to stick to null here at least for now. We can discuss null vs optional outside of this PR, but given that using null is significantly more prevalent in Gradle I think its best to stick with that

Reply:
Leaving `null` here for now is fine.  I'm putting this in the same mental box as the `private static` discussion we had a few weeks ago - we should discuss this in an internal thread and come to a resolution on how we want to go.  It would be great to enable a (weak) warning for this.  I feel like it's related to the work to upgrade the JDK we're building on - `Optional`s will only become more capable in later versions of Java and returning them vs `null`s will support more functional patterns.